File: frontend/src/api/labels.ts
/**
 * Labels API client - embedding-based sort and search for the
 * Labels verify tab.
 *
 * Underlying technique is still "similarity" (cosine distance on DINOv2
 * embeddings); the user-facing naming reflects the unit of work instead.
 *
 * Sort and search return an NDJSON event stream from the backend. Use
 * `sortStream` / `searchStream` if you want progress events; the plain
 * `sort` / `search` helpers drain the stream and return only the final
 * result.
 */

import { api } from "../lib/api-client";
import type {
  CohortsResponse,
  LabelsFilesParams,
  LabelsFilesResponse,
  LabelsProgress,
  LabelStatsResponse,
  SearchRequest,
  SearchResponse,
  SortRequest,
  SortResponse,
} from "./types";

export type LabelsProgressPhase =
  | "detections"
  | "load"
  | "sort"
  | "neighbors";

export interface LabelsProgressEvent {
  type: "progress";
  phase: LabelsProgressPhase;
  done: number;
  total: number;
}

interface ResultEvent<T> {
  type: "result";
  // The result payload is spread alongside `type`, matching the backend
  // event shape: `{"type":"result", ...SortResponse}`.
  [key: string]: unknown extends T ? unknown : T[keyof T] | "result";
}

interface ErrorEvent {
  type: "error";
  message: string;
}

type StreamEvent<T> =
  | LabelsProgressEvent
  | ResultEvent<T>
  | ErrorEvent;

/**
 * Stream NDJSON from a backend endpoint, calling `onProgress` for each
 * progress event and resolving with the final result payload.
 *
 * Pass `body=undefined` to make the request a GET (used by the cohorts
 * endpoint which has no body). Otherwise the body is JSON-encoded and
 * the request method is POST.
 *
 * Uses fetch + getReader rather than the shared `api` helper because
 * `api.post` parses the whole response body as JSON; here we need
 * line-by-line parsing.
 */
async function streamNdjson<T>(
  url: string,
  body: unknown,
  onProgress: (e: LabelsProgressEvent) => void,
  signal?: AbortSignal,
): Promise<T> {
  const isPost = body !== undefined;
  const init: RequestInit = {
    method: isPost ? "POST" : "GET",
    signal,
  };
  if (isPost) {
    init.headers = { "Content-Type": "application/json" };
    init.body = JSON.stringify(body);
  }
  const response = await fetch(url, init);

  if (!response.ok) {
    // Drain the body so the connection is reusable.
    const text = await response.text().catch(() => "");
    throw new Error(text || `HTTP ${response.status}`);
  }
  if (!response.body) {
    throw new Error("Response has no body");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let result: T | null = null;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });

    // Newline-delimited JSON. Keep the trailing partial line in the
    // buffer until the next read fills it.
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      const event = JSON.parse(trimmed) as StreamEvent<T>;
      if (event.type === "progress") {
        onProgress(event);
      } else if (event.type === "error") {
        throw new Error(event.message);
      } else if (event.type === "result") {
        const { type: _t, ...rest } = event as { type: "result" } & T;
        result = rest as unknown as T;
      }
    }
  }

  if (result === null) {
    throw new Error("No result event received");
  }
  return result;
}

export const labelsApi = {
  /**
   * Sort labels by visual similarity. Drains the NDJSON stream
   * and returns only the final result. Use `sortStream` if you also
   * need progress events.
   */
  sort: async (
    projectId: string,
    body: SortRequest,
  ): Promise<SortResponse> => {
    return labelsApi.sortStream(projectId, body, () => {});
  },

  sortStream: async (
    projectId: string,
    body: SortRequest,
    onProgress: (e: LabelsProgressEvent) => void,
    signal?: AbortSignal,
  ): Promise<SortResponse> => {
    return streamNdjson<SortResponse>(
      `/api/projects/${projectId}/labels/sort`,
      body,
      onProgress,
      signal,
    );
  },

  /** Find labels similar to an anchor. */
  search: async (
    projectId: string,
    body: SearchRequest,
  ): Promise<SearchResponse> => {
    return labelsApi.searchStream(projectId, body, () => {});
  },

  searchStream: async (
    projectId: string,
    body: SearchRequest,
    onProgress: (e: LabelsProgressEvent) => void,
    signal?: AbortSignal,
  ): Promise<SearchResponse> => {
    return streamNdjson<SearchResponse>(
      `/api/projects/${projectId}/labels/search`,
      body,
      onProgress,
      signal,
    );
  },

  /** Count detections in a confidence range that could appear in the
   * grid after an embedding backfill but currently cannot (embeddable,
   * yet no embedding for the current model). Drives the "unprocessed
   * detections" banner. */
  unprocessedCount: async (
    projectId: string,
    minConfidence: number,
    maxConfidence: number,
  ): Promise<{ count: number }> => {
    return api.get<{ count: number }>(
      `/api/projects/${projectId}/labels/unprocessed-count` +
        `?min_confidence=${minConfidence}&max_confidence=${maxConfidence}`,
    );
  },

  /**
   * The project's files for the Files tab, one item per file rather
   * than one per box. `empty` narrows to the files where nothing passed
   * the current floor ("show_only") or where something did ("hide").
   */
  files: async (
    projectId: string,
    params: LabelsFilesParams = {},
  ): Promise<LabelsFilesResponse> => {
    const sp = new URLSearchParams();
    if (params.site_ids?.length) sp.set("site_ids", params.site_ids.join(","));
    if (params.date_from) sp.set("date_from", params.date_from);
    if (params.date_to) sp.set("date_to", params.date_to);
    if (params.verification) sp.set("verification", params.verification);
    if (params.empty) sp.set("empty", params.empty);
    if (params.flagged && params.flagged !== "all") {
      sp.set("flagged", params.flagged);
    }
    if (params.favorited && params.favorited !== "all") {
      sp.set("favorited", params.favorited);
    }
    if (params.labels?.length) sp.set("labels", params.labels.join(","));
    if (params.min_confidence !== undefined) {
      sp.set("min_confidence", String(params.min_confidence));
    }
    if (params.max_confidence !== undefined) {
      sp.set("max_confidence", String(params.max_confidence));
    }
    if (params.min_label_confidence !== undefined) {
      sp.set("min_label_confidence", String(params.min_label_confidence));
    }
    if (params.max_label_confidence !== undefined) {
      sp.set("max_label_confidence", String(params.max_label_confidence));
    }
    if (params.sort) sp.set("sort", params.sort);
    if (params.seed !== undefined && params.seed !== null) {
      sp.set("seed", String(params.seed));
    }
    if (params.skip) sp.set("skip", String(params.skip));
    if (params.limit) sp.set("limit", String(params.limit));
    if (params.find) sp.set("find", params.find);
    return api.get<LabelsFilesResponse>(
      `/api/projects/${projectId}/labels/files?${sp.toString()}`,
    );
  },

  /**
   * Progress for the Labels page, counted in photos so both tabs can
   * show the same number.
   */
  progress: async (
    projectId: string,
    params: {
      site_ids?: string[];
      date_from?: string;
      date_to?: string;
      min_confidence?: number;
    } = {},
  ): Promise<LabelsProgress> => {
    const sp = new URLSearchParams();
    if (params.site_ids?.length) sp.set("site_ids", params.site_ids.join(","));
    if (params.date_from) sp.set("date_from", params.date_from);
    if (params.date_to) sp.set("date_to", params.date_to);
    if (params.min_confidence !== undefined)
      sp.set("min_confidence", String(params.min_confidence));
    return api.get<LabelsProgress>(
      `/api/projects/${projectId}/labels/progress?${sp.toString()}`,
    );
  },

  /** Get embedding coverage stats for a project. */
  stats: async (projectId: string): Promise<LabelStatsResponse> => {
    return api.get<LabelStatsResponse>(
      `/api/projects/${projectId}/labels/stats`,
    );
  },

  /**
   * Cohort grouping for the promotion review panel. Drains the NDJSON
   * stream and returns only the final result. Use `cohortsStream` to
   * surface progress.
   */
  cohorts: async (projectId: string): Promise<CohortsResponse> => {
    return labelsApi.cohortsStream(projectId, () => {});
  },

  cohortsStream: async (
    projectId: string,
    onProgress: (e: LabelsProgressEvent) => void,
    signal?: AbortSignal,
  ): Promise<CohortsResponse> => {
    // GET request — no body. The endpoint reads `min_count` and
    // `max_cohorts` from query params; defaults (5 / 20) are baked into
    // the panel's expected workload, so callers don't override them.
    return streamNdjson<CohortsResponse>(
      `/api/projects/${projectId}/labels/cohorts`,
      undefined,
      onProgress,
      signal,
    );
  },
};

File: backend/app/api/routers/labels.py
"""
Labels API router.

Drives the Labels verify tab: embedding-based sort, nearest-neighbor
search, and embedding coverage stats. Heavy computation is delegated to
ml/inference/similarity_script.py via subprocess (no numpy/faiss needed here).
The "similarity" name on internal modules reflects the underlying technique;
user-facing surfaces are named after the unit of work, the label.
"""

from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import StreamingResponse
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.api.crud import file as file_crud
from app.api.schemas.label import (
    EmptyFilter,
    LabelsFileItem,
    LabelsFilesResponse,
    LabelsFilesSort,
    LabelsProgress,
    LabelStatsResponse,
    LabelsVerification,
    SearchRequest,
    SortRequest,
)
from app.core.confidence import (
    DEFAULT_CLASSIFICATION_GATE,
    MD_OUTPUT_CONFIDENCE_THRESHOLD,
)
from app.db.base import get_db
from app.ml.label_exclusion import threshold_or_verified
from app.models import Deployment, Detection, DetectionEmbedding, File, Project
from app.services.label_service import (
    stream_cohorts_async,
    stream_search_async,
    stream_sort_async,
)
from app.utils.datetime_serialization import set_active_project_timezone

router = APIRouter(prefix="/api/projects", tags=["labels"])


def _parse_dt(value: str | None, field: str) -> datetime | None:
    """Parse an ISO date query param, or raise 422.

    ``datetime.fromisoformat`` raises ``ValueError``, which reaches the
    user as a 500 on a query they can fix themselves. The statistics
    router solves this the same way; the two stay separate rather than
    shared because that one parses to ``date`` and these endpoints need
    ``datetime``, and one helper serving both would need a flag.
    """
    if not value:
        return None
    try:
        return datetime.fromisoformat(value)
    except ValueError as err:
        raise HTTPException(
            status_code=422, detail=f"Invalid ISO date for {field}: {value}"
        ) from err


def _set_project_tz(db: Session, project_id: str) -> None:
    """Activate the project's timezone in the request context."""
    tz = db.query(Project.timezone).filter(Project.id == project_id).scalar()
    if tz:
        set_active_project_timezone(tz)


@router.post("/{project_id}/labels/sort")
async def sort_detections(
    project_id: str,
    body: SortRequest,
    request: Request,
    db: Session = Depends(get_db),
):
    """Sort detections by visual similarity (greedy nearest-neighbor chain).

    Returns an `application/x-ndjson` event stream from the worker
    subprocess: progress lines while loading and computing, then a
    final `{"type":"result", ...}` line whose payload matches the
    legacy SortResponse shape. The frontend renders a progress bar
    from progress events and uses the result for the grid. When the
    client disconnects mid-stream (refresh, tab close, navigation)
    the subprocess is killed promptly, so the browser's per-host
    connection slot is freed and the next page load isn't queued.
    """
    _set_project_tz(db, project_id)
    try:
        stream = stream_sort_async(request, project_id, body, db)
    except FileNotFoundError as e:
        raise HTTPException(status_code=503, detail=str(e)) from None
    return StreamingResponse(stream, media_type="application/x-ndjson")


@router.post("/{project_id}/labels/search")
async def search_similar(
    project_id: str,
    body: SearchRequest,
    request: Request,
    db: Session = Depends(get_db),
):
    """Find detections visually similar to an anchor detection.

    Same NDJSON event-stream shape as the sort endpoint; the final
    `result` payload matches SearchResponse.
    """
    _set_project_tz(db, project_id)
    try:
        stream = stream_search_async(request, project_id, body, db)
    except FileNotFoundError as e:
        raise HTTPException(status_code=503, detail=str(e)) from None
    return StreamingResponse(stream, media_type="application/x-ndjson")


@router.get("/{project_id}/labels/cohorts")
async def cohorts(
    project_id: str,
    request: Request,
    min_count: int = Query(8, ge=1, le=1000),
    max_cohorts: int = Query(200, ge=1, le=1000),
    db: Session = Depends(get_db),
):
    """Group descendant-promotion suggestions for the Edit-step review panel.

    Returns an `application/x-ndjson` event stream identical in shape to
    the sort endpoint: progress lines during the FAISS + neighbour walk,
    then a final `{"type":"result", "cohorts":[...]}` line whose payload
    matches CohortsResponse. The panel renders a skeleton from progress
    events and uses the result to render its cards. Cohorts span the
    whole project, no filters apply.
    """
    _set_project_tz(db, project_id)
    try:
        stream = stream_cohorts_async(
            request, project_id, min_count, max_cohorts, db
        )
    except FileNotFoundError as e:
        raise HTTPException(status_code=503, detail=str(e)) from None
    return StreamingResponse(stream, media_type="application/x-ndjson")


@router.get("/{project_id}/labels/unprocessed-count")
def get_unprocessed_count(
    project_id: str,
    min_confidence: float = MD_OUTPUT_CONFIDENCE_THRESHOLD,
    max_confidence: float = 1.0,
    db: Session = Depends(get_db),
) -> dict:
    """Count detections in a confidence range that could appear in the
    labels grid after an embedding backfill, but currently cannot: they
    are embeddable (bbox, on an embeddable surface) yet have no
    embedding for the project's current embedding model.

    Drives the grid's "unprocessed detections" banner. Purely
    data-driven, so it stays correct for projects whose deployments
    were analysed under different classification gates: whatever the
    historical gate was, an un-embedded detection shows up here.
    """
    from sqlalchemy import and_, exists, or_

    project = db.query(Project).filter(Project.id == project_id).first()
    embedding_model_id = project.embedding_model_id if project else None
    if not embedding_model_id:
        # No embedding model: the grid cannot show anything either way,
        # and the existing "Embed now" callout owns that situation.
        return {"count": 0}

    has_bbox = Detection.bbox_x.isnot(None)
    on_embeddable_surface = or_(
        File.file_type == "image",
        and_(
            File.file_type == "video",
            Detection.frame_number == File.best_frame_number,
        ),
    )
    has_embedding = exists().where(
        and_(
            DetectionEmbedding.detection_id == Detection.id,
            DetectionEmbedding.embedding_model_id == embedding_model_id,
        )
    )

    count = (
        db.query(func.count(Detection.id))
        .join(File, File.id == Detection.file_id)
        .join(Deployment, Deployment.id == File.deployment_id)
        .filter(Deployment.project_id == project_id)
        .filter(has_bbox)
        .filter(on_embeddable_surface)
        .filter(~has_embedding)
        .filter(Detection.confidence >= min_confidence)
        .filter(Detection.confidence <= max_confidence)
        .scalar()
    ) or 0

    return {"count": int(count)}


@router.get(
    "/{project_id}/labels/files",
    response_model=LabelsFilesResponse,
)
async def get_labels_files(
    project_id: str,
    site_ids: str | None = Query(None, description="Comma-separated site IDs"),
    date_from: str | None = Query(None, description="ISO date (YYYY-MM-DD)"),
    date_to: str | None = Query(None, description="ISO date (YYYY-MM-DD)"),
    verification: LabelsVerification | None = Query(None),
    flagged: str | None = Query(None, pattern="^(all|flagged|not_flagged)$"),
    favorited: str | None = Query(
        None, pattern="^(all|favorited|not_favorited)$"
    ),
    empty: EmptyFilter = Query("all"),
    labels: str | None = Query(
        None, description="Comma-separated label taxonomy IDs"
    ),
    min_confidence: float | None = Query(None, ge=0.0, le=1.0),
    max_confidence: float | None = Query(None, ge=0.0, le=1.0),
    min_label_confidence: float | None = Query(None, ge=0.0, le=1.0),
    max_label_confidence: float | None = Query(None, ge=0.0, le=1.0),
    sort: LabelsFilesSort = Query("path"),
    seed: int | None = Query(None, description="Required for sort=random"),
    skip: int = Query(0, ge=0),
    limit: int = Query(48, ge=1, le=200),
    find: str | None = Query(
        None,
        description="File id whose position in the full ordering to report",
    ),
    db: Session = Depends(get_db),
):
    """The project's files for the Files tab, one item per file.

    The other half of the Labels page. The crop grid shows every
    detection above the floor, one card per box; this shows the files
    themselves, so a whole photo can be judged and a photo the detector
    dismissed is still reachable. `empty` narrows to the files where
    nothing passed the floor (`show_only`) or where something did
    (`hide`); the floor is the project's counting threshold, always, so
    the two halves partition the project and a transient slider can
    never redefine what "empty" means.

    `async def` because `captured_at_local` is an observational datetime
    and its serializer reads the project timezone from a ContextVar set
    here; a sync endpoint would set it in a threadpool where the
    serializer cannot see it (DEVELOPERS.md "Datetime conventions").
    """
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    if sort == "random" and seed is None:
        raise HTTPException(
            status_code=400, detail="sort=random requires a seed"
        )
    _set_project_tz(db, project_id)

    # "Empty" is defined by the project's counting threshold alone; the
    # confidence slider on Files is clamped there and never drags the
    # floor down (that is a Detections-only affordance, where the weak
    # boxes can actually be shown as cards). A below-floor
    # `min_confidence` from a stale URL is a no-op in the CRUD.
    floor = project.counting_threshold
    total, files, find_index = file_crud.get_labels_files(
        db,
        project_id,
        floor=floor,
        empty=empty,
        site_ids=site_ids.split(",") if site_ids else None,
        date_from=_parse_dt(date_from, "date_from"),
        date_to=_parse_dt(date_to, "date_to"),
        verification=verification,
        flagged=flagged,
        favorited=favorited,
        labels=labels.split(",") if labels else None,
        min_confidence=min_confidence,
        max_confidence=max_confidence,
        min_label_confidence=min_label_confidence,
        max_label_confidence=max_label_confidence,
        sort=sort,
        seed=seed,
        skip=skip,
        limit=limit,
        find=find,
    )
    items = [LabelsFileItem.model_validate(f) for f in files]
    if sort == "events" and files:
        # One batch lookup for the page, matching the sort's own key (a
        # file's first event). The grid draws divider rows from it.
        from app.models.event import event_files as ef

        rows = db.execute(
            select(ef.c.file_id, ef.c.event_id)
            .where(ef.c.file_id.in_([f.id for f in files]))
            # Deterministic pick for a file in several events, matching
            # the sort key's own ORDER BY event_id LIMIT 1.
            .order_by(ef.c.event_id)
        ).all()
        first_event: dict[str, str] = {}
        for file_id, event_id in rows:
            first_event.setdefault(file_id, event_id)
        for item in items:
            item.event_id = first_event.get(item.id)
    return LabelsFilesResponse(
        total=total,
        floor=floor,
        items=items,
        find_index=find_index,
    )


@router.get(
    "/{project_id}/labels/progress",
    response_model=LabelsProgress,
)
async def get_labels_progress(
    project_id: str,
    site_ids: str | None = Query(None, description="Comma-separated site IDs"),
    date_from: str | None = Query(None, description="ISO date (YYYY-MM-DD)"),
    date_to: str | None = Query(None, description="ISO date (YYYY-MM-DD)"),
    min_confidence: float | None = Query(None, ge=0.0, le=1.0),
    db: Session = Depends(get_db),
):
    """Progress for the Labels page, counted in labels.

    One number for the whole page rather than one per tab: a detection
    above the threshold is a label to check, and a file with nothing
    above it is a label too, "nothing here". The total is the number of
    cards across both tabs, so 100% means every one has been looked at.

    The dashboard reads this same endpoint, so its bar and the page's
    pill can never disagree.

    `async def` for consistency with the other endpoints here; it returns
    no observational datetimes of its own.
    """
    if not db.query(Project.id).filter(Project.id == project_id).first():
        raise HTTPException(status_code=404, detail="Project not found")

    counts = file_crud.get_label_progress(
        db,
        project_id,
        site_ids=site_ids.split(",") if site_ids else None,
        date_from=_parse_dt(date_from, "date_from"),
        date_to=_parse_dt(date_to, "date_to"),
        min_confidence=min_confidence,
    )
    return LabelsProgress(
        total_labels=counts.total,
        verified_labels=counts.verified,
        crop_labels=counts.crop_labels,
        crop_labels_verified=counts.crop_labels_verified,
        empty_labels=counts.empty_labels,
        empty_labels_verified=counts.empty_labels_verified,
        files=counts.files,
        files_verified=counts.files_verified,
    )


@router.get(
    "/{project_id}/labels/stats",
    response_model=LabelStatsResponse,
)
def get_label_stats(
    project_id: str,
    db: Session = Depends(get_db),
):
    """Get embedding coverage stats for a project."""
    # Get embedding model from project first (needed to filter embedded count)
    project = db.query(Project).filter(Project.id == project_id).first()
    embedding_model_id = project.embedding_model_id if project else None

    # Embeddability gate. A detection is embeddable when:
    #  - it has a bbox (event-level labels are bbox-less and have
    #    no crop to embed), AND
    #  - it sits on a pixel surface the embedding worker can read:
    #    images embed unconditionally; video detections only embed when
    #    they sit on the parent video's best frame (matches
    #    `build_embedding_input` in embedding_utils.py).
    # Non-embeddable detections are invisible to the Labels grid
    # and similarity search anyway, so we leave them out of the
    # "missing embeddings" count — otherwise the banner would chase a
    # population that `/embed-now` is deliberately designed to skip.
    from sqlalchemy import and_, exists, or_
    has_bbox = Detection.bbox_x.isnot(None)
    on_embeddable_surface = or_(
        File.file_type == "image",
        and_(
            File.file_type == "video",
            Detection.frame_number == File.best_frame_number,
        ),
    )
    embeddable_clause = and_(has_bbox, on_embeddable_surface)

    # Only detections at or above the classification gate (or verified)
    # are SUPPOSED to be embedded — MegaDetector runs untresholded, so
    # the below-gate tail is deliberately unprocessed and must not
    # count as "missing" here (the grid's separate unprocessed-range
    # banner owns that population, with its own backfill action).
    gate = (
        project.classification_gate
        if project
        else DEFAULT_CLASSIFICATION_GATE
    )
    above_gate_or_verified = threshold_or_verified(gate)

    total = (
        db.query(func.count(Detection.id))
        .join(File, File.id == Detection.file_id)
        .join(Deployment, Deployment.id == File.deployment_id)
        .filter(Deployment.project_id == project_id)
        .filter(embeddable_clause)
        .filter(above_gate_or_verified)
        .scalar()
    ) or 0

    # The verification progress pill does NOT come from here. It reads
    # `verified_detections` / `total_detections` off
    # `/api/events/verification-stats`, which counts the whole reviewable
    # population rather than only the embedded part of it, so the Labels
    # pill and the dashboard bar agree. Do not add a percentage to this
    # endpoint: requiring an embedding would under-count, because the
    # event sort renders detections that have none.
    has_embedding = exists().where(
        DetectionEmbedding.detection_id == Detection.id
    )
    verified = (
        db.query(func.count(Detection.id))
        .join(File, File.id == Detection.file_id)
        .join(Deployment, Deployment.id == File.deployment_id)
        .filter(Deployment.project_id == project_id)
        .filter(has_embedding)
        # Same scope as `total` above, or a signed-off file's rejected
        # weak boxes (verified, embedded) count as reviewed work here.
        .filter(above_gate_or_verified)
        .filter(Detection.verified == True)  # noqa: E712
        .scalar()
    ) or 0

    # Detections with embeddings for the current model
    if embedding_model_id:
        embedded = (
            db.query(func.count(func.distinct(DetectionEmbedding.detection_id)))
            .join(Detection, Detection.id == DetectionEmbedding.detection_id)
            .join(File, File.id == Detection.file_id)
            .join(Deployment, Deployment.id == File.deployment_id)
            .filter(Deployment.project_id == project_id)
            .filter(embeddable_clause)
            .filter(DetectionEmbedding.embedding_model_id == embedding_model_id)
            .scalar()
        ) or 0
    else:
        embedded = 0

    # Get dimension from first embedding
    dim_row = (
        db.query(DetectionEmbedding.dimension)
        .join(Detection, Detection.id == DetectionEmbedding.detection_id)
        .join(File, File.id == Detection.file_id)
        .join(Deployment, Deployment.id == File.deployment_id)
        .filter(Deployment.project_id == project_id)
        .limit(1)
        .first()
    )

    return LabelStatsResponse(
        total_detections=total,
        verified_detections=verified,
        embedded_detections=embedded,
        missing_embeddings=total - embedded,
        embedding_model_id=embedding_model_id,
        embedding_dimension=dim_row[0] if dim_row else None,
    )
